# polyprotein_stats
A small tool to test and visualize protein embeddings and amino acid proportions.

Currently deployed on [streamlit.io](https://share.streamlit.io/leonfrench/polyprotein_stats/main/protein_embeddings_app.py).

Given a set of proteins, this tool seeks to answer these questions:

* is there a enrichment of one or two amino acids? 
* can amino acid proportions, plus sequence length discriminate the input proteins from the rest of the proteome?
* can the averaged per residue embeddings from a deep neural network discriminate the input proteins from the rest of the proteome?

Embeddings for the human proteins is from the ProtT5 embedder at full precision (prottrans_t5_xl_u50 model) that were 
generated by Christian Dallago and Burkhard Rost ([Zenodo](https://zenodo.org/record/5047020#.YgQ_fe7MI-R)). Also, see 'ProtTrans: Towards Cracking the Language of Lifeâ€™s Code Through Self-Supervised Learning' by Elnaggar et al. [biorxiv preprint](https://www.biorxiv.org/content/10.1101/2020.07.12.199554v3)

Embeddings for the human genes are from the Geneformer model weights from the [Hugging Face](https://huggingface.co/ctheodoris/Geneformer) repository. Full details on Geneformer by Theodoris et al. is available at [Nature](https://www.nature.com/articles/s41586-023-06139-9). 

Embeddings for the human genes are from the scGPT model weights from the [Github](https://github.com/bowang-lab/scGPT) repository (model: Pretrained on 33 million normal human cells). Full details on scGPT by Cui et al. is available as a [preprint](https://www.biorxiv.org/content/10.1101/2023.04.30.538439v2). 

The default gene list is from the [Lindbohm et al.](https://alz-journals.onlinelibrary.wiley.com/doi/10.1002/alz.12419) study of cognitive decline and risk of dementia.

Note - when deploying on streamlit.io use python version 3.10 (not the default!)